scripts/rv_expression_association/prepare_inputs.py

```diff
@@ -0,0 +1,212 @@
+    )
+
+    ####################
+    ### kinship file ###
+    ####################
+
+    ## read in GRM (genotype relationship matrix; kinship matrix)
+    K = pd.read_csv(kinship_file, index_col=0)
+    K.index = K.index.astype('str')
+    assert all(K.columns == K.index)  # symmetric matrix, donors x donors
+
+    K = xr.DataArray(
+        K.values,
+        dims=['sample_0', 'sample_1'],
+        coords={'sample_0': K.columns, 'sample_1': K.index},
+    )
+    K = K.sortby('sample_0').sortby('sample_1')
+
+    #####################
+    ### genotype files ##
+    #####################
+
+    # read in genotype file (plink format)
+    # bed
+    with to_path(genotype_file_bed).open('rb') as handle:
+        data = handle.readlines()
+    with open('temp.bed', 'wb') as handle:
+        handle.writelines(data)
+    # bim
+    with to_path(genotype_file_bim).open('rb') as handle:
+        data = handle.readlines()
+    with open('temp.bim', 'wb') as handle:
+        handle.writelines(data)
+    # fam
+    with to_path(genotype_file_fam).open('rb') as handle:
+        data = handle.readlines()
+    with open('temp.fam', 'wb') as handle:
+        handle.writelines(data)
+    # read
+    G = read_plink1_bin('temp.bed')
+
+    #################################
+    ### sample mapping file (SMF) ###
+    #################################
+
+    # this file will map different IDs (and OneK1K ID to CPG ID)
+    sample_mapping = pd.read_csv(sample_mapping_file, sep='\t')
+
+    ## samples with expression data
+    donors_onek1k = sample_mapping['OneK1K_ID'].unique()
+    donors_onek1k.sort()
+    donors_exprs = sorted(
+        set(list(phenotype.sample.values)).intersection(donors_onek1k)
+    )
+    logging.info(
+        'Number of unique donors with expression data: {}'.format(len(donors_exprs))
+    )
+
+    ## samples with genotype data
+    donors_cpg = sample_mapping['InternalID'].unique()
+    donors_cpg.sort()
+    donors_geno = sorted(set(list(G.sample.values)).intersection(donors_cpg))
+    logging.info(
+        'Number of unique donors with genotype data: {}'.format(len(donors_geno))
+    )
+
+    ## samples with both (can this be done in one step?)
+    sample_mapping1 = sample_mapping.loc[sample_mapping['OneK1K_ID'].isin(donors_exprs)]
+    sample_mapping_both = sample_mapping1.loc[
+        sample_mapping1['InternalID'].isin(donors_geno)
+    ]
+    donors_e = sample_mapping_both['OneK1K_ID'].unique()
+    donors_g = sample_mapping_both['InternalID'].unique()
+    assert len(donors_e) == len(donors_g)
+
+    logging.info('Number of unique common donors: {}'.format(len(donors_g)))
+
+    ## samples in kinship
+    donors_e_short = [re.sub('.*_', '', donor) for donor in donors_e]
+    donors_k = sorted(set(list(K.sample_0.values)).intersection(donors_e_short))
+
+    ######################################
+    ### subset files to common samples ###
+    ######################################
+
+    ###############
```

Comment:
You might want to consider using [marker regions](https://code.visualstudio.com/docs/editor/codebasics#_folding) instead.

Reply:
i like the idea but am finding the VSCode documentation a bit obscure still :( do you have a good example in python by any chance? Sorry for being slow

Reply:
It's fairly simple to use -- here's an example for Python: https://github.com/populationgenomics/tob-wgs/blob/7fddfcc795672f5d6f66e43ce9bec12467412119/scripts/eqtl_hail_batch/launch_eqtl_spearman.py#L75

Reply:
sorry, perfect, thanks!